File: 백준/Silver/2512. 예산/예산.py
N = int(input());
budget = list(map(int, input().split()));
M = int(input());
s= sum(budget)
min_value = 1
max_value = 100000
answer=0
if s<=M:
    print(max(budget))
else:
    while min_value<=max_value:
        mid=(min_value+max_value)//2
        budget_sum=0
        for value in budget:
            if value<mid:
                budget_sum+=value
            else:
                budget_sum+=mid
        if budget_sum>M:
            max_value=mid-1
        elif budget_sum<=M:
            min_value=mid+1
            answer=mid
    print(answer)

File: 백준/Silver/2512. 예산/예산.java
import java.io.*;
import java.util.*;

public class Main {

    private static int N;
    private static int[] arr;
    private static int total;
    public static void main(String[] args) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        N = Integer.parseInt(br.readLine());
        arr = new int[N];
        StringTokenizer st = new StringTokenizer(br.readLine());
        int s = -1;
        int e =0;

        for(int i=0; i<N; i++){
            arr[i] = Integer.parseInt(st.nextToken());
            if(arr[i]>e){
                e=arr[i];
            }
        }

        total = Integer.parseInt(br.readLine());

        e+=1;

        while(s+1<e){
            int mid = (s+e)/2;
            if(check(mid)) s=mid;
            else e=mid;
        }
        System.out.println(s);

    }

    public static boolean check(int target){
        int sum=0;
        for(int budget : arr){
            if(budget>target){
                sum+=target;
            }else{
                sum+=budget;
            }
        }

        return sum<=total;
    }
}
